### PR TITLE
Nim 1.0

### PR DESCRIPTION
**Pull request by juancarlospaco**
- https://github.com/nim-lang/Nim/commit/5b43c98897ee7eb9f8ded8ceb7623f6caa23bace
- Compile-Time optional PackedJSON for Performance (only changes `import json`) https://github.com/Araq/packedjson#packedjson

A new update cycle :slightly_smiling_face:

### CODE REVIEW

File: nimwc.nim [hunk 1]
```diff
@@ -109,14 +111,14 @@ const
     when defined(useStdoutAsStdmsg): " -d:useStdoutAsStdmsg", # Use Std Out as Std Msg
     when defined(nimOldShiftRight):  " -d:nimOldShiftRight",  # http://forum.nim-lang.org/t/4891#30600
     when defined(nimOldCaseObjects): " -d:nimOldCaseObjects", # old case switch
-    when defined(nimBinaryStdFiles): " -d:nimBinaryStdFiles", # stdin/stdout old binary open
+    when defined(nimDontSetUtf8CodePage): " -d:nimDontSetUtf8CodePage", # stdin/stdout Not UTF-8
```
Comment (ThomasTJdev) [inline nimwc.nim]: Is this also for Linux? Changelog states:
```
On Windows console applications the code-page is set at program startup to UTF-8. Use the new switch -d:nimDontSetUtf8CodePage to disable this feature.
```

File: nimwc.nim [hunk 2]
```diff
@@ -109,14 +111,14 @@ const
...
   ].deduplicate.join  ## Checking for known compile options and returning them as a space separated string at Compile-Time. See README.md for explanation of the options.
 
   nimwc_version =
     try:
       filterIt("nimwc.nimble".readFile.splitLines, it.substr(0, 6) == "version")[0].split("= ")[1].normalize.replace("\"", "") ## Get NimWC Version at Compile-Time.
     except:
-      "5.5.2"  ## Set NimWC Version at Compile-Time, if reading from file fails.
+      "5.5.5"  ## Set NimWC Version at Compile-Time, if reading from file fails.
```
Comment (ThomasTJdev) [inline nimwc.nim]: Can we use the new [`NimblePkgVersion`](https://github.com/nim-lang/nimble/blob/master/changelog.markdown#0110---22092019)?

```
The NimblePkgVersion is now defined so you can easily get the package version in your source code (example).
```

https://github.com/nim-lang/nimble/blob/4a2aaa07d/tests/nimbleVersionDefine/src/nimbleVersionDefine.nim

File: nimwcpkg/nimwc_main.nim [hunk 3]
```diff
@@ -1,13 +1,3 @@
...
-#        (c) Copyright 2019 Thomas Toftgaard Jarløv
-#        Look at LICENSE for more info.
-#        All rights reserved.
-import
-  asyncdispatch, bcrypt, cgi, jester, json, macros, os, osproc, logging, otp,
-  parsecfg, random, re, sequtils, strutils, times, datetime2human,
-  base32, streams, encodings, nativesockets, libravatar, contra,
-  oswalkdir
-
 import
```
Comment (ThomasTJdev) [inline nimwcpkg/nimwc_main.nim]: Maybe a comment about that the imports are now placed in `nimwc_main.nim.cfg`.

File: nimwc.nim (hunk 1 above)
Comment (juancarlospaco) [reply nimwc.nim]: Fixed.

File: nimwc.nim (hunk 2 above)
Comment (juancarlospaco) [reply nimwc.nim]: Fixed. I didnt know about it :grey_exclamation:

File: nimwcpkg/resources/administration/createdb.nim [hunk 4]
```diff
@@ -243,6 +222,6 @@ proc backupDb*(dbname: string,
     info("Database backup: Fail - " & filename)
 
 
-proc vacuumDb*(db: DbConn): bool {.inline.} =
+template vacuumDb*(db: DbConn): bool =
```
Comment (ThomasTJdev) [inline nimwcpkg/resources/administration/createdb.nim]: Question: Why change to template?

File: nimwcpkg/resources/administration/createdb.nim.cfg [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import: os
+import: parsecfg
+import: strutils
+import: contra
+import: osproc
+
+
+@if postgres:
```
Comment (ThomasTJdev) [inline nimwcpkg/resources/administration/createdb.nim.cfg]: Nice one 🎉

File: nimwcpkg/resources/password/password_generate.nim [hunk 6]
```diff
@@ -1,14 +1,14 @@
 import md5, bcrypt, contra, ../password/salt_generate
 
 
-proc makeSessionKey*(): string {.inline.} =
+template makeSessionKey*(): string =
```
Comment (ThomasTJdev) [inline nimwcpkg/resources/password/password_generate.nim]: Question: Why make them templates?

File: nimwcpkg/resources/administration/createdb.nim (hunk 4 above)
Comment (juancarlospaco) [reply nimwcpkg/resources/administration/createdb.nim]: `proc  {.inline.}` is almost the same as `template` anyways.
For performance reasons mainly, but if the function is tiny is the same.

**Comment by juancarlospaco** [comment]
I want to move static const very long strings out of code, 
because it makes it look like is huge when is not, and static const strings we rarely touch,
so Nim offers the following options:

- `*.nimf` is for code with logic, for frontend, not good for static strings, is still more code.
- `include "file.nim"` is for code with logic, for backend, not good for static strings, is still more code, can end up with Nim logic in there too.
- `staticRead("file.foo")` is for generic strings, need 1 file per `const`, lots of files, but can work Ok too.
- `const foo {.strdefine.} = ""` is for `const` strings, is not code, cant have Nim logic which is good, can have a default value, 1 file for all `const`, can only have compile-time `if`, `elif`, `else`.

**Comment by ThomasTJdev** [comment]
Nice work @juancarlospaco!

I was thinking about adding a small **License** block in our README - or a explanation inside the License.md (?). The PPL is new to many, and can push people away. Something like your responses on the old issues:

```
[TLDR PPL](https://tldrlegal.com/license/peer-production-license)
The Peer Production License (PPL) is derived from the Creative Commons ‘Attribution-NonCommercial-ShareAlike’ license. However unlike CC-BY-NC-SA, PPL demands reciprocation (whether monetary or otherwise) from non-contributing commercial entities (aka Big Corporations). This means, that a mega-corporation cannot just take all the code, rename it, and sell it without contributing back.

The Peer Production License is a reciprocity-based license by which commons are freely accessible to those who contribute to create them, while third party non-contributing entities profiting from these commons can be charged a license fee to sustain the project in the long run.
```
Besides the above, a explation, bullet list or FAQ for:
* If you are a independent developer, you may: Use it xx
* If you own a software company, and the company helps with PR's, you may: Use it xx
* If you own a company, which wants to use NimWC to provide free website to it users, you may: Use it and pay
* May I use NimWC (without editing) as my webpage, where i sell products?
* May I use NimWC (without editing) with my own plugin, where I sell the service my plugin provides?

**Comment by juancarlospaco** [comment]
The same file layout I did with the `constants.nim` I will try to reorder the internal modules,
the reason is to reduce the I/O unneeded repetition when reading the configs again and again,
for performance and DRY, besides constants look nice now.

**Comment by juancarlospaco** [comment]
From inspecting the generated code of the `*.nimf`:

- They wont need `# result = ""` inside the function anymore, that was when `string` was `nil`.
- Can be `func`.
- Can be `{.inline.}`.

**Comment by ThomasTJdev** [comment]
> From inspecting the generated code of the `*.nimf`:
> 
> * They wont need `# result = ""` inside the function anymore, that was when `string` was `nil`.
> * Can be `func`.
> * Can be `{.inline.}`.
> 
> I am beautifying the `*.nimf`.

Wow - great catch! :+1:

**Comment by ThomasTJdev** [comment]
Idea: style.css and style_custom.css should be merged into 1 file avoid 2 file request (also JS)?

**Comment by juancarlospaco** [comment]
`--backuplogs  Compressed full backup of all old unused rotated Logs (TarGz).`
Makes all `./logs/*.log` :arrow_right: `logs-ISODATETIME.tar.gz`. :slightly_smiling_face:

**Comment by juancarlospaco** [comment]
Too much run-time `encodeUrl("foo bar")`, I will try to move those into `constants.nim` (compile-time).

:thinking:

**Comment by juancarlospaco** [comment]
```nim
get "/sitemap.xml":
    writeFile("public/sitemap.xml", genSitemap())
    sendFile("public/sitemap.xml")
```

Here `writeFile("public/sitemap.xml", genSitemap())` needs to be run and re-write the file on each request :grey_question: 
Or we can wrap it on a `once:` so it writes the file once per run :grey_question: 
:thinking:

**Comment by ThomasTJdev** [comment]
> `routes.nim`
> 
> ```nim
> get "/sitemap.xml":
>     writeFile("public/sitemap.xml", genSitemap())
>     sendFile("public/sitemap.xml")
> ```
> 
> Here `writeFile("public/sitemap.xml", genSitemap())` needs to be run and re-write the file on each request 
> Or we can wrap it on a `once:` so it writes the file once per run 
> 
> If needs to run on each request, it needs to be written into a file 
> Can just `resp genSitemap()`

Yeah, I know. When a user is publishing a lot of new blogpost, it's nice to have the sitemap.xml updated, but the downside is the IO. We can make a update-`proc` which is runned on:
1) startup
2) blog creating
3) blog updating
4) page creating
5) page updating

This way the user does not need to remember to update the sitemap or restart NimWC.

**Comment by juancarlospaco** [comment]
Ok, then it makes sense to update it, my question was if it can be compile-time constant.

**Comment by juancarlospaco** [comment]
Hit a bug on Jester, can you run this code on your local to check if its failing :grey_question: 

```nim
import jester

routes:
  get "/":
    resp """
    <form name="login" action="/dologin" method="post">
      <input type="text" id="username" name="username" value="somevalue">
      <input type="submit"/>
    </form>
    """

  post "/dologin":
    assert @"username" == "somevalue", @"username"
```

**Comment by juancarlospaco** [comment]
@ThomasTJdev Review and let me know any comments or push the change to the branch directly. :slightly_smiling_face:

File: nimwc.nimble [hunk 7]
```diff
@@ -10,16 +10,17 @@ installDirs   = @["config", "nimwcpkg", "plugins", "public"]
...
-requires "bcrypt >= 0.2.1"
-requires "contra >= 0.2.0"
-requires "datetime2human >= 0.2.2"
-requires "firejail >= 0.5.0"
-requires "jester >= 0.4.1"
-requires "libravatar >= 0.4.0"
-requires "otp >= 0.1.1"
-requires "recaptcha >= 1.0.2"
-requires "webp >= 0.2.0"
+requires "nim >= 1.0.2"
```
[the remark this answers is not in the document]
Comment (juancarlospaco) [reply nimwc.nimble]: https://nim-lang.org/blog/2019/10/23/version-102-released.html

[the remark this answers is not in the document]
Comment (ThomasTJdev) [reply nimwc.nimble]: Yeah I know, but do we need 1.0.2? If just install 1.0.0 due to package managers, they can't install with nimble.

File: nimwcpkg/resources/administration/connectdb.nim [hunk 8]
```diff
@@ -1,36 +1,29 @@
...
     assert db_name.len > 0, "db_name must not be empty string."
-    assert db_host.len > 2, "db_host must not be empty string."
-  else: discard existsOrCreateDir(db_folder)  # Creating folder
-
-  var db {.inject.} =
-    when defined(postgres):
-      db_postgres.open(connection=db_host, user=db_user, password=db_pass, database=db_name)
-    else:
-      db_sqlite.open(db_host, "", "", "")
+    var db {.inject.} = db_postgres.open(connection=db_host, user=db_user, password=db_pass, database=db_name)
```
[the remark this answers is not in the document]
Comment (juancarlospaco) [reply nimwcpkg/resources/administration/connectdb.nim]: https://nim-lang.github.io/Nim/manual.html#templates-hygiene-in-templates

File: nimwc.nimble (hunk 7 above)
[the remark this answers is not in the document]
Comment (juancarlospaco) [reply nimwc.nimble]: Thats not our Bug. :P 
Too many Distros if we want to fix them all.

[the remark this answers is not in the document]
Comment (juancarlospaco) [reply nimwc.nimble]: I just got an issue of someone using Centos, was using GCC 4, GCC is like 9, options wont even exist for that decade.
🤷‍♀️

File: .gitignore [hunk 9]
```diff
@@ -54,3 +54,5 @@ TODO.md
 nimwc
 nimwc_main
 nimwcpkg/nimwc_main
+
+localhost
```
[the remark this answers is not in the document]
Comment (juancarlospaco) [reply .gitignore]: Fixed.

File: nimwc.nimble (hunk 7 above)
[the remark this answers is not in the document]
Comment (ThomasTJdev) [reply nimwc.nimble]: Okay!

**Comment by ThomasTJdev** [comment]
Great @juancarlospaco :smiley: !! What are missing before a merge?

**Comment by juancarlospaco** [comment]
I fixed the backup of logs to put `.tar.gz` on `/logs/`.

**Comment by ThomasTJdev** [comment]
New release is ready: https://github.com/ThomasTJdev/nim_websitecreator/releases

I'll start on the plugins, they all need an upgrade due to the refactoring.

**Comment by juancarlospaco** [comment]
Dont forget the Live Demo too.

I know that it breaks plugins, but was a refactor to reorder stuff that was kind of long delayed, 
now the folder structure kind of makes sense, and is easy to navigate for new people, looks better.

**Pull request merged**